### ISSUE TITLE
Support different encodings

### ISSUE THREAD

**Issue by Artilaina** [ForkIssues/TrackerWin#75]
All the files under my project are in ANSI encoding but it seems that Fork is unable to show them correctly:
![imagen](https://user-images.githubusercontent.com/25099519/49748272-866e9e80-fca5-11e8-8b0d-17c853c16cb5.png)

Besides, when it comes to stage some block with special characters like "ñ" or "ü" it shows the following error:

> error: patch failed: Fuentes/Cliente/Formularios/frmMensajesLog.frm:1184
> error: Fuentes/Cliente/Formularios/frmMensajesLog.frm: patch does not apply

**Comment by LuizzTrevisan** [comment]
UP

**Comment by DanPristupov** [comment]
Does staging for such files work anywhere else?

**Comment by Artilaina** [comment]
Sure. I'm still using GIT GUI without issues.

**Comment by DanPristupov** [comment]
@Artilaina thanks. Does it determine encoding automatically?

**Comment by Artilaina** [comment]
No, I think that it uses utf-8 by design but it can be changed to any other encoding, both by global config or by repository:
![git_gui](https://user-images.githubusercontent.com/25099519/51371407-5fa05500-1afa-11e9-8791-322f712a9ba2.png)

**Comment by DanPristupov** [comment]
Is it possible in reality that a repository has files with different encodings?
I'm trying to understand the problem better.

**Comment by Artilaina** [comment]
I'm almost sure that it can't be possible: the project couldn't compile files with different encoding other that ANSI
Certanly, inside the project folder there are logs and other files in UTF-8 but all of them are excluded with .gitignore so it's supposed that they are outside the repository control, isn't?

**Comment by DanPristupov** [comment]
> I'm almost sure that it can't be possible: the project couldn't compile files with different encoding other that ANSI

If I remember correctly, Visual Studio (C#) used to propose to save file as unicode when I typed a non-ascii character. So, it depends on compiler.

OK, I have enough information for now. I don't have any ETA though. Support for different encodings may require very large changes.

**Comment by Artilaina** [comment]
Visual Studio? C#? 😄 😭 It's an ancient VB6 project with proC services, thats why we are tied to ansi.
Thanks anyway. I will wait for it.

**Comment by LuizzTrevisan** [comment]
in my case I use Delphi for development, all files are in ansi, in the preview changes the accents are not shown.

![image](https://user-images.githubusercontent.com/2445313/51381893-7647ab80-1afc-11e9-8489-5ba8e5f7b8e1.png)

**Comment by hypersw** [comment]
A repo can easily have files with different encodings. Even within the same solution, but then it's common to have more than one isolated solution in a large repo. Plus test data, that would be in different encodings intentionally.

UTF-8 has redundancy in its coding which ensures that you can tell apart UTF-8 and ANSI in most cases. Visual Studio would read a file as UTF-8 and fall back to ANSI if any invalid characters are encountered. Adding some magic for sensing UTF-16, plus looking at the BOMs as the first step, pretty much sums this up and works well for most real-life files.

**Comment by mloskot** [comment]
@DanPristupov  https://github.com/ForkIssues/TrackerWin/issues/75#issuecomment-455459261
> Is it possible in reality that a repository has files with different encodings?

Yes, it is.

For example, Windows users often will want to have this in `.gitattributes`

```
*.rc working-tree-encoding=UTF-16LE-BOM -text
```

I have experienced that staging such files with Fork 'breaks' the encoding, while staging with `git` directly in Git Bash works fine.

I have not submitted any bug report yet because it's still unclear to me what are the actual steps to reproduce it. I can't understand why it happens, what Fork does to those files, what commands Fork runs.

**Comment by noxthot** [comment]
I would also really like to see this feature. My department is also stuck to ANSI files in our repositories.

We also occasionally suffer from the "patch does not apply" error due to encoding reasons. Staging the whole file still works then, but staging of specific blocks/lines within a file fails.

**Comment by anthroid** [comment]
Same issue. In my case, the files are created by the IDE with ANSI/Windows 1252 encoding by default and it is not possible to change them or use UTF-8 (also not an option to use a different IDE since it's tied to hardware). The same files are displayed correctly, detected automatically, and can be committed without any issue in Sublime Merge, VS Code, Atom, Notepad++, command line, etc.

**Comment by KazzmanK** [comment]
Upvoting!

Autodetecting encoding is a bit challenging, so there is no need to do too much magic here.
I have windows-1251 files that sometimes treated as ISO-8859-5
So I personally will be happy to have encoding switcher like we have in GIT GUI
![image](https://user-images.githubusercontent.com/13328513/185227028-ec1340da-4bb1-4658-bae0-adee43eeaaa0.png)

![image](https://user-images.githubusercontent.com/13328513/185227169-7081eb25-5699-4088-90fc-6d06606e8e06.png)

This will make me more confident of what changes are actually done to my files and not, afraid too much about file corruptions.
Will it?

btw, changes between different encodings can`t be done via patches, i.e. partial staging. Is it expected ?

**Comment by pfaustinom1** [comment]
Any news about this? I'm facing the same problem.

**Comment by waterborns** [comment]
Up :(
